consensus/src/detail/tests/double_spend_tests.rs

```diff
@@ -0,0 +1,314 @@
+        //
+        // Genesis -> b1 -> b2 where
+        //
+        // +--Block-1--------------+
+        // |                       |
+        // | +-------tx-1--------+ |
+        // | |input = genesis    | |
+        // | +-------------------+ |
+        // +-----------------------+
+        //
+        // +--Block-2--------------+
+        // |                       |
+        // | +-------tx-1--------+ |
+        // | |input = genesis    | |
+        // | +-------------------+ |
+        // +-----------------------+
+
+        let mut consensus = setup_consensus();
+        let receiver = random_address(&consensus.chain_config);
+
+        let prev_block_tx_id =
+            consensus.chain_config.genesis_block().transactions().get(0).unwrap().get_id();
+
+        // Create first tx
+        let first_tx = Transaction::new(
+            0,
+            vec![TxInput::new(
+                OutPointSourceId::Transaction(prev_block_tx_id.clone()),
+                0,
+                random_witness(),
+            )],
+            vec![TxOutput::new(Amount::from_atoms(12345678912345), receiver.clone())],
+            0,
+        )
+        .expect(ERR_CREATE_TX_FAIL);
+
+        // Create tx that pointing to the previous tx
+        let first_block = Block::new(
+            vec![first_tx],
+            Some(Id::new(
+                &consensus.chain_config.genesis_block().get_id().get(),
+            )),
+            time::get() as u32,
+            ConsensusData::None,
+        )
+        .expect(ERR_CREATE_BLOCK_FAIL);
+        let first_block_id = first_block.get_id();
+        assert!(consensus.process_block(first_block, BlockSource::Local).is_ok());
+        assert_eq!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND),
+            Some(first_block_id.clone())
+        );
+        // Create second tx
+        let second_tx = Transaction::new(
+            12345,
+            vec![TxInput::new(
+                OutPointSourceId::Transaction(prev_block_tx_id),
+                0,
+                random_witness(),
+            )],
+            vec![TxOutput::new(Amount::from_atoms(12345678912345), receiver)],
+            0,
+        )
+        .expect(ERR_CREATE_TX_FAIL);
+
+        // Create tx that pointing to the previous tx
+        let second_block = Block::new(
+            vec![second_tx],
+            Some(first_block_id.clone()),
+            time::get() as u32,
+            ConsensusData::None,
+        )
+        .expect(ERR_CREATE_BLOCK_FAIL);
+        assert!(consensus.process_block(second_block, BlockSource::Local).is_err());
+        assert_eq!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND)
+                .expect(ERR_STORAGE_FAIL),
+            first_block_id
+        );
+    });
+}
```

Comment:
For all these tests, if there are forks, please show them in the little ASCII art you're using. Better use something like bitcoin does in feature_blocks.py

Reply:
In this file, no forks so far.